docs/development.md

```diff
@@ -2,136 +2,103 @@
-configuration:
-
-- **C/C++** - [`clang-format`](https://clang.llvm.org/docs/ClangFormat.html)
-  with the [Mozilla C++
-  style](https://firefox-source-docs.mozilla.org/code-quality/coding-style/coding_style_cpp.html)
-- **Python** - [Black](https://github.com/psf/black) with an 80 char line limit
-- **Markdown** - wrap to 80 characters where possible (not enforced)
-- **CMake** - TBD
-- **Bash** - TBD
-
-## Checking out the repo
-
-For now, many locations rely on the code being located at
-`/usr/local/code/faasm` (note the lowercase "f"), with the latter two 
-directories _owned_ by the current user. This is annoying and something we will 
-fix in future. 
-
-You can either set this directory up directly, or just symlink it.
-
-Assuming you've checked out this code somewhere, you'll need to make sure
-submodules are up to date:
+Most development can be done using the containerised CLI. To get everything set
+up, you need to run:
 
 ```bash
-git submodule update --init
-```
+# Clone this repo
+git clone https://github.com/faasm/faasm
+cd faasm
 
-## Basic local machine set-up
+# Update submodules
+git submodule update --init
 
-Most of the local set-up is scripted with Ansible, but you need to have Python 3
-and [Ansible](https://www.ansible.com/) set up in advance.
+# Start the CLI
+./bin/cli.sh
+```
 
-The easiest way to do this is as follows:
+To build and run the tests, you can then run the following inside the container:
 
 ```bash
-# Python stuff and other dependencies
-sudo apt install python3-dev python3-pip python3-venv libcairo2-dev libtinfo5
-sudo pip3 install -U pip
+# Build the tests and the codegen targets
+inv dev.cmake
+inv dev.cc tests
+inv dev.cc codegen_func
+inv dev.cc codegen_shared_obj
 
-# Ansible
-sudo pip install -U ansible
+# Upload the Python functions
+inv upload.user python --local-copy --py
 
-# Faasm python env 
-source workon.sh
-pip install -r faasmcli/requirements.txt
+# Run codegen (this may take a while the first time it's run)
+inv codegen.local
 
-# Faasm CLI
-pip install -e faasmcli/
+# Set up cgroup
+./bin/cgroup.sh
 
-# Faasm playbook
-cd ansible
-ansible-playbook local_dev.yml --ask-become-pass
+# Run the tests
+tests
 ```
 
-If you want to tweak things yourself, look inside the `local_dev.yml` playbook
-to see what's required.
-
-### Protobuf and gRPC
+### Tooling - editors, IDEs etc.
 
-Faasm depends on Protobuf and gRPC which are set up with the Ansible `grpc` role
-found [here](ansible/roles/grpc/tasks/main.yml). However, if you want to intsall
-these yourself or use an existing version, make sure you use the full 
-[gRPC C++ build](https://grpc.io/docs/languages/cpp/quickstart/).
+Several directories (including the project root) are mounted as volumes in the
+CLI container (see the [CLI `docker-compose.yml`](../docker-compose-cli.yml)). 
+This means you can edit the files directly on your host and compile them inside 
+the container.
```

Comment:
There's a problem with this approach, and it's the old [root/uid problem](https://github.com/moby/moby/issues/3124#issuecomment-104937065), you can also see this in their [docs](https://docs.docker.com/compose/compose-file/#configs). In particular, from outside the container, all the artifacts generated from _within_ the container, will be owned by root. I would personally use the same `uid:gid` as the user invoking the script, hence from outside the container we would not see a difference in the generated builds.

We could either modify the underlying `cli.dockerfile` or `docker-compose-cli.yml`, I think the latter is cleaner as we can easily query  parameters from the command line. I'd rather centralize my suggestions in this comment than in the respective files.

_After working for an hour on it_
I have some ideas on how to do this, and would be willing to do a PR myself. It's not trivial though. We could discuss this personally.

Reply:
Yes I'm aware of this, and although it can occasionally be inconvenient, how much of a problem is it in this case?

The directories used for volumes are either implicitly read only (the sources), or created specifically for mounting (i.e. `build` or `container/*`). Neither `build` nor `container` are checked in, and can be owned completely by the container, so this shouldn't cause any conflicts. The only issue I can see here is if you create new source files from within the container, which could become a problem.

Reply:
As we discussed, I agree on your point, and I see that _practically_ this is not an issue. However, I consider that installing a piece of software and, as an artifact, ending up with `root`-owned (noun phrase) directories in your home folder to be far from desirable.

I also understand that this is somewhat unrelated to this PR, and the container-based-dev will keep evolving, so I move the discussion to #328.